src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs

```diff
@@ -0,0 +1,1306 @@
+            }
+
+            return path;
+        }
+
+        private static Exception GetExceptionIfPartUriInvalid(Uri partUri, out string partUriString)
+        {
+            partUriString = String.Empty;
+
+            if (partUri == null)
+                return new ArgumentNullException("partUri");
+
+            Exception argumentException = null;
+
+            argumentException = GetExceptionIfAbsoluteUri(partUri);
+            if (argumentException != null)
+                return argumentException;
+
+            string partName = GetStringForPartUriFromAnyUri(partUri);
+
+            //We need to make sure that the URI passed to us is not just "/"
+            //"/" is a valid relative uri, but is not a valid partname
+            if (partName == String.Empty)
+                return new ArgumentException(SR.PartUriIsEmpty);
+
+            if (partName[0] != '/')
+                return new ArgumentException(SR.PartUriShouldStartWithForwardSlash);
+
+            argumentException = GetExceptionIfPartNameStartsWithTwoSlashes(partName);
+            if (argumentException != null)
+                return argumentException;
+
+            argumentException = GetExceptionIfPartNameEndsWithSlash(partName);
+            if (argumentException != null)
+                return argumentException;
+
+            argumentException = GetExceptionIfFragmentPresent(partName);
+            if (argumentException != null)
+                return argumentException;
+
+            //We test if the URI is wellformed and refined.
+            //The relative URI that was passed to us may not be correctly escaped and so we test that.
+            //Also there might be navigation "/../" present in the URI which we need to detect.
+            string wellFormedPartName =
+                new Uri(s_defaultUri, partName).GetComponents(UriComponents.Path |
+                                                             UriComponents.KeepDelimiter, UriFormat.UriEscaped);
+
+            //Note - For Relative Uris the output of ToString() and OriginalString property 
+            //are the same as per the current implementation of System.Uri
+            //Need to use OriginalString property or ToString() here as we are want to 
+            //validate that the input uri given to us was valid in the first place.
+            //We do not want to use GetComponents in this case as it might lead to
+            //performing escaping or unescaping as per the UriFormat enum value and that 
+            //may alter the string that the user created the Uri with and we may not be able 
+            //to verify the uri correctly.
+            //We perform the comparison in a case-insensitive manner, as at this point,
+            //only escaped hex digits (A-F) might vary in casing.
+            if (String.CompareOrdinal(partUri.OriginalString.ToUpperInvariant(), wellFormedPartName.ToUpperInvariant()) != 0)
```

Comment:
Could this just do a ordinal-ignorecase comparison rather than allocating new strings with ToUpperInvariant?